addons/mail/models/mail_mail.py

```diff
@@ -325,7 +337,7 @@ def _send(self, auto_commit=False, raise_exception=False, smtp_session=None):
                 mail = self.browse(mail_id)
                 if mail.state != 'outgoing':
                     if mail.state != 'exception' and mail.auto_delete:
-                        mail.sudo().unlink()
+                        mail.sudo().state = 'todelete'
```

Comment:
Actually this code is strange ... I think it is not the job of _send to try to be an auto vacuum. Can't we just remove that code ? Seems everything should be managed based on state: a mail is sent and its state is updated to 'todelete' or 'sent' (depending on auto delete flag), or is set to 'exception'. Don't see why this could should be necessary now. It could be necessary in earlier version of the outgoing gateway but I think it can be removed now (in that case would make a specific commit with explanations).

Reply:
This is useful when the mail are set as cancelled in the mail composer (see screenshot)

Or we should move the "auto delete" logic in the mail composer, but I prefer as it is now

![screenshot_spect](https://user-images.githubusercontent.com/54316226/161713166-95a2c266-7aae-42d9-828e-3dc454e16233.png)

Reply:
Mmmh ... why would we remove them ? Those are canceled and there is an issue to solve somehow (or admin can remove them). Moreover the cron takes only outgoing emails, so they are not unlinked, unless I misunderstood something.

Reply:
E.G. if an email is not valid / missing, the mail composer will set the state as "cancel"

So for me it make sens to remove those mails

Or maybe we set directly "todelete" if the mail composer if auto_delete is True ?

Reply:
That's not what the current code does. As most emails are sent through the queue, domain is set to outgoing (see process_email_queue). Therefore they are not unlinked as they will not go through _send, which makes sense. No need to try to send canceled emails.

Changing this can be done, but is not the purpose of this task as we may have valid use case when we want to keep canceled emails.

Reply:
Indeed, from the "CRON pov which call only outgoing email", I understand :) 

I changed that in a separate commit, and adapt the test in the next commit :)